aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/compute/strategy/CreateResourcesThenCreateNodes.java

```diff
@@ -166,51 +218,89 @@ public boolean apply(KeyPair input) {
             }
          });
-         for (String securityGroupName : options.getGroups()) {
-            checkState(Iterables.contains(securityGroupNames, securityGroupName), "Cannot find security group with name " + securityGroupName + ". \nSecurity groups available are: \n" + Iterables.toString(securityGroupNames)); // {
+      }
+
+      // check whether a security group with the same inbound ports is available
+      Map<Integer, Integer> portRanges = getPortRangesFromList(inboundPorts);
+      List<String> portRangesAsString = Lists.newArrayList(Iterables.transform(portRanges.entrySet(), new Function<Map.Entry<Integer, Integer>, String>() {
+         @Override
+         public String apply(Map.Entry<Integer, Integer> input) {
+            return String.format(PORT_RANGE_FORMAT, input.getKey(), input.getValue());
          }
-      } else if (options.getInboundPorts().length > 0) {
-         String name = namingConvention.create().sharedNameForGroup(group);
-         SecurityGroupRequest securityGroupRequest = api.securityGroupApi().create(location.getId(),
-                 CreateSecurityGroupOptions.Builder.securityGroupName(name).vpcId(options.getVpcId()));
-         // add rules
-         Map<Integer, Integer> portRanges = getPortRangesFromList(options.getInboundPorts());
-         for (Map.Entry<Integer, Integer> portRange : portRanges.entrySet()) {
-            String range = portRange.getKey() + "/" + portRange.getValue();
-            // TODO makes protocol and source CIDR configurable?
-            api.securityGroupApi().addInboundRule(
-                    location.getId(),
-                    securityGroupRequest.getSecurityGroupId(),
-                    IpProtocol.TCP,
-                    range,
-                    INTERNET);
+      }));
+
+      for (SecurityGroup securityGroup : api.securityGroupApi().list(regionId).concat()) {
+         List<Permission> permissions = api.securityGroupApi().get(regionId, securityGroup.id());
+         if (permissions.size() == portRangesAsString.size()) {
+            for (Permission permission : permissions) {
+               if (permission.ipProtocol() == IpProtocol.TCP && permission.sourceCidrIp().equals(INTERNET)) {
+                  portRangesAsString.remove(permission.portRange());
+               }
+            }
+            if (portRangesAsString.isEmpty()) return Optional.of(securityGroup);
          }
-         api.tagApi().add(location.getId(), securityGroupRequest.getSecurityGroupId(), SECURITYGROUP,
-                 TagOptions.Builder.tag(1, "owner", "jclouds"));
-         securityGroupId = securityGroupRequest.getSecurityGroupId();
       }
-      return securityGroupId;
+      return Optional.absent();
+   }
+
+   private Optional<VSwitch> tryFindVSwitchInRegion(String regionId, String vSwitchId) {
+      if (Strings.isNullOrEmpty(vSwitchId)) {
+         return api.vSwitchApi().list(regionId).concat().firstMatch(Predicates.<VSwitch>notNull());
```

Comment:
Is it correct to connect our nodes to _any_ vSwitch we found, if none was provided? We could be just connecting things to existing isolated networks that have a very concrete purpose. I wouldn't do this and always create a vSwitch for jclouds if none was provided.